node/modules/services.go

```diff
@@ -73,14 +76,50 @@ func RunBlockSync(h host.Host, svc *blocksync.BlockSyncService) {
 	h.SetStreamHandler(blocksync.BlockSyncProtocolID, svc.HandleStream)
 }
 
-func HandleIncomingBlocks(mctx helpers.MetricsCtx, lc fx.Lifecycle, ps *pubsub.PubSub, s *chain.Syncer, bserv dtypes.ChainBlockService, chain *store.ChainStore, stmgr *stmgr.StateManager, h host.Host, nn dtypes.NetworkName) {
-	ctx := helpers.LifecycleCtx(mctx, lc)
+func waitForSync(stmgr *stmgr.StateManager, blocks int, subscribe func()) {
+	nearsync := uint64(blocks) * uint64(build.BlockDelaySecs) * uint64(time.Second) //nolint
 
-	blocksub, err := ps.Subscribe(build.BlocksTopic(nn)) //nolint
-	if err != nil {
-		panic(err)
+	// early check, are we synced at start up?
+	ts := stmgr.ChainStore().GetHeaviestTipSet()
+	timestamp := ts.MinTimestamp()
+	now := uint64(build.Clock.Now().UnixNano())
+	if timestamp > now-nearsync {
+		subscribe()
+		return
 	}
 
+	// we are not synced, subscribe to head changes and wait for sync
+	subscribed := false
+	stmgr.ChainStore().SubscribeHeadChanges(func(rev, app []*types.TipSet) error {
+		if subscribed {
+			return nil
+		}
```

Comment:
We should figure out how to remove this handler, instead of keeping it and ignoring events (e.g. by returning a special error and handling it in chain/store/store.go:reorgWorker where we call notifees)

Reply:
yeah, that's a good idea.

Reply:
done!